lib/Dialect/Torch/Transforms/RefineTypes.cpp

```diff
@@ -759,6 +759,44 @@ static void fillInSizesGivenSizesList(ValueKnowledge &knowledge, Value sizes) {
   }
 }
 
+static void fillInSizesForBinaryBroadcastingOp(ValueKnowledge &lhs,
+                                               ValueKnowledge &rhs,
+                                               ValueKnowledge &knowledge) {
+  if (lhs.hasSizes && rhs.hasSizes) {
+    knowledge.hasSizes = true;
+    knowledge.sizes.resize(std::max(lhs.sizes.size(), rhs.sizes.size()),
+                           kUnknownSize);
+
+    int64_t resultRank = knowledge.sizes.size();
+    auto increaseRankToResultRank =
+        [&](const std::vector<int64_t> &sizes) -> std::vector<int64_t> {
+      int offset = resultRank - sizes.size();
+      std::vector<int64_t> newSizes(std::max(offset, 0), 1);
+      newSizes.insert(newSizes.end(), sizes.begin(), sizes.end());
+      return newSizes;
+    };
+
+    std::vector<int64_t> rankAdjustedSizesLhs =
+        increaseRankToResultRank(lhs.sizes);
+    std::vector<int64_t> rankAdjustedSizesRhs =
+        increaseRankToResultRank(rhs.sizes);
+
+    for (int64_t i = 0; i < resultRank; i++) {
+      int64_t lhsDimSize = rankAdjustedSizesLhs[i];
+      int64_t rhsDimSize = rankAdjustedSizesRhs[i];
+      // Dynamic shape can't be decided at compilation.
+      if (lhsDimSize == kUnknownSize || rhsDimSize == kUnknownSize)
+        continue;
+
+      // Incompatible broadcasting shape.
+      if (lhsDimSize != rhsDimSize && lhsDimSize != 1 && rhsDimSize != 1)
+        continue;
```

Comment:
If we've statically determined that the input tensors are of the wrong shape, is it still correct to return a shape as the final knowledge, or should the knowledge be set to the pessimistic value?

I'm trying to understand how to interpret the following expected invariant:

https://github.com/llvm/torch-mlir/blob/34562b42b1295ea5074627bfb845bab6f311737b/lib/Dialect/Torch/Transforms/RefineTypes.cpp#L117-L118

It seems to me that when we were not explicitly handling the static size analysis done here, there existed scenarios where the invariant would be broken. In particular, in the case where the input tensors had static non-broadcastable sizes, we were still wrongly claiming to know the exact rank of the result. However, it would not be useful to just return the pessimistic result for everything. Ignoring static analysis forced us to assume that the inputs are always well formed for broadcasting.

However, now that we are performing static analysis, we can be a bit more strict about the invariant. What do you think?

Reply:
yea, that's a good point. Thanks for the thorough thought on this.